org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ChangeSignatureHandler.java

```diff
@@ -0,0 +1,173 @@
+
+public class ChangeSignatureHandler {
+
+	public static String CHANGE_SIGNATURE_ANNOTATION_ID = "java.refactor.ChangeSignature";
+
+	public static class MethodParameter {
+		public String type;
+		public String name;
+		public String defaultValue;
+		public int originalIndex;
+
+		public MethodParameter(String type, String name, String defaultValue, int originalIndex) {
+			this.type = type;
+			this.name = name;
+			this.defaultValue = defaultValue;
+			this.originalIndex = originalIndex;
+		}
+	}
+
+	public static class MethodException {
+		public String type;
+		public String typeHandleIdentifier;
+
+		public MethodException(String type, String typeHandleIdentifier) {
+			this.type = type;
+			this.typeHandleIdentifier = typeHandleIdentifier;
+		}
+	}
+
+	public static Refactoring getChangeSignatureRefactoring(CodeActionParams params, IMethod method, boolean isDelegate, String methodName, String modifier, String returnType, List<MethodParameter> parameters,
+			List<MethodException> exceptions) {
+		ICompilationUnit cu = JDTUtils.resolveCompilationUnit(params.getTextDocument().getUri());
+		if (cu == null) {
+			return null;
+		}
+		IType primaryType = cu.findPrimaryType();
+		if (primaryType == null) {
+			return null;
+		}
+		try {
+			ChangeSignatureProcessor processor = new ChangeSignatureProcessor(method);
+			processor.setNewMethodName(methodName);
+			processor.setVisibility(JdtFlags.getVisibilityCode(modifier));
+			processor.setNewReturnTypeName(returnType);
+			processor.setDelegateUpdating(isDelegate);
+			RefactoringStatus status = processor.checkInitialConditions(new NullProgressMonitor());
+			if (status.hasFatalError()) {
+				if (status.hasFatalError()) {
+					logFatalError(status);
+					return null;
+				}
+			}
+			List<ParameterInfo> parameterInfos = processor.getParameterInfos();
+			List<ParameterInfo> newParameterInfos = new ArrayList<>();
+			for (MethodParameter param : parameters) {
+				if (param.originalIndex != ParameterInfo.INDEX_FOR_ADDED && param.originalIndex < parameterInfos.size()) {
+					ParameterInfo info = parameterInfos.get(param.originalIndex);
+					info.setNewTypeName(param.type);
+					info.setNewName(param.name);
+					newParameterInfos.add(info);
+				} else {
+					newParameterInfos.add(ParameterInfo.createInfoForAddedParameter(param.type, param.name, param.defaultValue));
+				}
+			}
+			parameterInfos.clear();
+			parameterInfos.addAll(newParameterInfos);
+			List<ExceptionInfo> exceptionInfos = processor.getExceptionInfos();
+			List<ExceptionInfo> newExceptionInfos = new ArrayList<>();
+			for (MethodException exception : exceptions) {
+				if (exception.typeHandleIdentifier != null) {
+					IJavaElement element = JavaCore.create(exception.typeHandleIdentifier);
+					if (element instanceof IType type) {
+						newExceptionInfos.add(ExceptionInfo.createInfoForAddedException(type));
+					}
+				} else {
+					IType type = cu.getJavaProject().findType(exception.type);
```

Comment:
Can we do some context-aware search first here? For example:

1. Check if the exception type appears in the import declarations.
2. If it's `Exception`, just simply treat it as `java.lang.Exception`

Reply:
https://github.com/eclipse/eclipse.jdt.ls/pull/2497/commits/eda6e385c8cbed7004736564f9601009af54932c introduces several handlings for this.